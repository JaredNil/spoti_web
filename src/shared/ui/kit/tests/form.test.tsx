import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { useForm } from 'react-hook-form'

import {
	Form,
	FormField,
	FormItem,
	FormLabel,
	FormControl,
	FormDescription,
	FormMessage,
	useFormField,
} from '../form'

const TestForm = ({
	onSubmit = jest.fn(),
}: {
	onSubmit?: (data: any) => void
}) => {
	const methods = useForm<{ name: string }>({ defaultValues: { name: '' } })
	return (
		<Form {...methods}>
			<form onSubmit={methods.handleSubmit(onSubmit)}>
				<FormField
					name="name"
					render={({ field }) => (
						<FormItem>
							<FormLabel>Name</FormLabel>
							<FormControl>
								<input {...field} data-testid="input" />
							</FormControl>
							<FormDescription>Your full name</FormDescription>
							<FormMessage />
						</FormItem>
					)}
				/>
				<button type="submit">submit</button>
			</form>
		</Form>
	)
}

describe('Form components', () => {
	it('FormItem наличие и работоспособность', () => {
		render(<FormItem className="custom">item</FormItem>)
		const div = screen.getByText('item')
		expect(div).toHaveAttribute('data-slot', 'form-item')
		expect(div).toHaveClass('grid')
	})

	it('FormDescription наличие и работоспособность', () => {
		render(<TestForm />)
		expect(screen.getByText('Your full name')).toBeInTheDocument()
	})

	it('FormMessage не работоспособность при отсутствии ошибок', () => {
		render(<TestForm />)
		expect(
			screen.queryByRole('paragraph', { name: /error/i })
		).not.toBeInTheDocument()
	})

	it('snapshot полной формы', () => {
		const { container } = render(<TestForm />)
		expect(container.firstChild).toMatchSnapshot()
	})
})
