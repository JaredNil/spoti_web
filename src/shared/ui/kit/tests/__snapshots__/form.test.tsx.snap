// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Form components snapshot полной формы 1`] = `
<form>
  <div
    class="grid gap-2"
    data-slot="form-item"
  >
    <label
      class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive"
      data-error="false"
      data-slot="form-label"
      for="«r3»-form-item"
    >
      Name
    </label>
    <input
      aria-describedby="«r3»-form-item-description"
      aria-invalid="false"
      data-slot="form-control"
      data-testid="input"
      id="«r3»-form-item"
      name="name"
      value=""
    />
    <p
      class="text-muted-foreground text-sm"
      data-slot="form-description"
      id="«r3»-form-item-description"
    >
      Your full name
    </p>
  </div>
  <button
    type="submit"
  >
    submit
  </button>
</form>
`;
